Github App Tech Refactor

sasi2312 opened the pull request:
Complete PR of App tech ticket

1. Once deployed in prod..pls add webhook secret in webhooks
2. Add that secret key in .env file named "WEBHOOK_SECRET"

KDwevedi commented:
@sasi2312  Can you add a list of changes you're pushing in this PR?
- implemented A
- fixed B

Like so ^

sasi2312 commented:
These are the changes made on this PR, @KDwevedi 

- Created seperate handler class for each webhook events
- Implemented secretkey verifications in webhooks
- Created dispatcher file to call respective eventhandlers
- Created new handlers for issuecomments , pull requests
- Implemented webhook event & dispatcher as generic manner
- Craeted DB adapter using singleton pattern 
- Implemented the initialization of db adapter class as singleton instance
- Craeted Github adapter & clustering github related functions into this class
- Made changes on Ingestor file as per prod file
- Refactored some existing function in generic way

KDwevedi commented on app.py:
```diff
@@ -27,49 +30,45 @@
+        r = await GithubAdapter.get_github_data(code)
+
+        auth_token = (r)["access_token"]
+        headers = {
```
if `headers` is being re-defined here, let's get rid of the definition on on Line#36

KDwevedi commented on app.py:
```diff
@@ -27,49 +30,45 @@
-            auth_token = (r)["access_token"]
-            headers = {
-                "Authorization": f"Bearer {auth_token}"
+        r = await GithubAdapter.get_github_data(code)
```
Let's name this variable better
Something like `responseData`

KDwevedi commented on app.py:
```diff
@@ -27,49 +30,45 @@
+        headers = {
+            "Authorization": f"Bearer {auth_token}"
+        }
+        async with session.get("https://api.github.com/user", headers=headers) as user_response:
```
Let's wrap this in a function in Github Adaptor? `GetUser` ?

sasi2312 replied:
okay ... will change

KDwevedi commented on app.py:
```diff
@@ -81,24 +80,22 @@ async def comment_cleaner():
 async def fetch_github_issues_from_repo(owner, repo):
-    url = f"https://api.github.com/repos/{owner}/{repo}/issues"
+    try:
+        response = await GithubAdapter.fetch_github_issues_from_repo(owner,repo)
```
Let's just use camelCase for function definitions as well.

KDwevedi commented on app.py:
```diff
@@ -81,24 +80,22 @@ async def comment_cleaner():
+                
+    except Exception as e:
+        logger.info(e)
+        pass
```
why is 
```python
pass
```

here?

sasi2312 replied:
Im not sure to raise a exception in this place ot not so i used pass here @KDwevedi

KDwevedi commented on handlers/installation_handler.py:
```diff
@@ -0,0 +1,17 @@
+            if data.get("installation") and data["installation"].get("account"):
+                # if data["action"] not in ["deleted", "suspend"]:
+                await TicketEventHandler().updateInstallation(data.get("installation"))
+            pass
```
Why use `pass` here

sasi2312 replied:
Its just a placeholder here.

KDwevedi commented on handlers/pull_request_handler.py:
```diff
@@ -0,0 +1,38 @@
+            
+            save_data = supabase_client.add_data(pr_data,"pr_history")            
+           
+            pass
```
Why use `pass` here?

KDwevedi commented on supabasedatapipeline/github_profile_render/ingestor.py:
```diff
@@ -65,11 +67,166 @@ def update(self, data):
+                return None,None
+
+            #GET TEMPLATES
+            if milestone == 1:
```
This should be a check using truth values of badge related table columns

sasi2312 replied:
Im not clear with this comment @KDwevedi  .... pls provide detailed explanations

KDwevedi commented on utils/github_adapter.py:
```diff
@@ -0,0 +1,232 @@
+                    r = await response.json()                    
+                    return r
+                
+        except Exception as e:
```
Have we thought through exception handling and logging?
If, for example, this endpoint starts failing, we'll stop receiving github details of individuals registering through the discord flow.

We should probably raise an alert for that, for example on a discord webhook on the testing server/samagra server?

sasi2312 replied:
Im not clear about this comment... @KDwevedi  For now i raise an exception from except block

KDwevedi commented on utils/github_adapter.py:
```diff
@@ -0,0 +1,232 @@
+            return None
+            
+        
+    async def get_calssroom_data(assignment_id,page):
```
typo in function name
calssroom -> classroom

KDwevedi commented on handlers/issue_comment_handler.py:
```diff
@@ -0,0 +1,32 @@
+            save_data = supabase_client.add_data(comment_data,"ticket_comments")
+                     
+        except Exception as e:
+            logger.info(f"{datetime.now()}---jobs works")
```
we shouldn't be logging a success message if the function failed to save data to supabase

KDwevedi commented on utils/github_adapter.py:
```diff
@@ -0,0 +1,232 @@
+                    print(f'Response body: {response_text}',file=sys.stderr)
+    
+    
+    async def deleteComment(self, owner, repo, comment_id):
```
let's have consistent naming for all variables and functions adhering to the [style guide](https://peps.python.org/pep-0008/#function-and-variable-names)

KDwevedi reviewed (changes requested):
- the .get_instance() paradigm in SupabaseInterface will slowdown API interactions with database trying to instantiate and reinstantiate class instances. Let's simply convert that to being a module and import functions for use.
- Let's define exception handling better and introduce some observability.
- Let's not use `pass` as placeholder statements for anything. Many uses either have no functionality here or hide the fact that a crucial function has failed.
- Github Profile Rendering logic seems outdated.
  - It's referencing template names with old badge names (enthusiast/rising star)

sasi2312 commented:
> * the .get_instance() paradigm in SupabaseInterface will slowdown API interactions with database trying to instantiate and reinstantiate class instances. Let's simply convert that to being a module and import functions for use.
> * Let's define exception handling better and introduce some observability.
> * Let's not use `pass` as placeholder statements for anything. Many uses either have no functionality here or hide the fact that a crucial function has failed.
> * Github Profile Rendering logic seems outdated.
>   
>   * It's referencing template names with old badge names (enthusiast/rising star)

sasi2312 commented:
- [ Github Profile Rendering logic seems outdated.
It's referencing template names with old badge names (enthusiast/rising star)]
 @KDwevedi 

- Can u please give me the context  of above statement

KDwevedi commented:
> Actually the db adapter part is designed as per singleton pattern ... It returns single(same) instance for every call.Its not slow down any database related part

You're correct.

the singleton pattern is only 1.25 times slower here:
```python
import time

# Using a singleton class
class MySingletonClass:
    _instance = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(MySingletonClass, cls).__new__(cls, *args, **kwargs)
        return cls._instance

    def my_method(self):
        pass

# Timing the singleton class method calls
singleton_instance = MySingletonClass()
start = time.time()
for _ in range(1000000):
    singleton_instance.my_method()
end = time.time()
print(f"Singleton class method call time: {end - start} seconds")

# Using a module
def my_function():
    pass

# Timing the module function calls
start = time.time()
for _ in range(1000000):
    my_function()
end = time.time()
print(f"Module function call time: {end - start} seconds")
```
```bash
Singleton class method call time: 0.05184793472290039 seconds
Module function call time: 0.039575815200805664 seconds
```

While the class initialisation is 2-3x slower:
```python
import time

# Using a class
class MyClass:
    def my_method(self):
        pass

start = time.time()
for _ in range(1000000):
    instance = MyClass()
    instance.my_method()
end = time.time()
print(f"Class method call time: {end - start} seconds")

# Using a module
def my_function():
    pass

start = time.time()
for _ in range(1000000):
    my_function()
end = time.time()
print(f"Module function call time: {end - start} seconds")
```
```bash
Class method call time: 0.11432194709777832 seconds
Module function call time: 0.04015994071960449 seconds
```

I still feel like the singleton pattern is an unnecessary overhead, but @VedantKhairnar can decide if we want to migrate to a module.

Let's leave this be.

KDwevedi commented:
> Can u please give me the context of above statement

The latest community badges are *Bronze*, *Silver*, *Gold*, *Ruby*, and *Diamond* according to the [leaderboard](https://www.codeforgovtech.in/leaderboard/)

The templates for the *Bronze*, *Silver*, and *Gold* badges are stored in the  [github_profile_assets](https://supabase.com/dashboard/project/kcavhjwafgtoqkqbbqrd/storage/buckets/github_profile_assets) bucket in supabase storage. 

The badges for *Ruby*, and *Diamond* can be made by copying the *Gold* or *Silver* or *Bronze* slide [here](https://docs.google.com/presentation/d/1-bNYTyKrilnZ4I8PVub6MGhNsDmTI9sN) 
(I've given edit access to sasikumar.s@cogniverselabs.com) and using the [Ruby](https://github.com/Code4GovTech/dmp-documentation/tree/main/assets/certificate_assets/Ruby.png) and [Diamond](https://github.com/Code4GovTech/dmp-documentation/tree/main/assets/certificate_assets/Diamond.png) png files.

In the new logic for generating github profile images you should
- [ ] Get rid of references to deprecated badges like enthusiast/rising star
- [ ] update the logic to also handle milestone 4 (Ruby) and milestone 5 (Diamond), as per point thresholds described in the community leaderboard
- [ ] add templates for these two badges
- [ ] validate the github_profile_data materialized view in supabase to make sure that milestone 4 and 5 are being generated correctly for valid point totals



Python script to generate the badges for reference:
```python
def generateQR(border_size=1, ghUsername=''):
    qr = segno.make_qr(f"https://github-app.c4gt.samagra.io/verify/{ghUsername}")
    buffer = BytesIO()
    # Save the QR code to a buffer, adjusting the border size here
    qr.save(buffer, kind='png', scale=1, border=border_size)
    qrBytes = buffer.getvalue()
    return qrBytes

profileData = client.table("github_profile_data").select("*").execute()
count = 0

for data in profileData.data:
    if data["dpg_points"]+data["classroom_points"]<10:
        continue

    

    #SET DATA
    milestone = data["milestone"]
    prsRaised = data["prs_raised"]
    prsReviewed = data["prs_reviewed"]
    prsMerged = data["prs_merged"]
    dpgPoints = data["dpg_points"]+data["classroom_points"]
    githubUsername = data["github_username"]

    githubLogoBytes = client.storage.from_('github_profile_assets').download("GHLogoSS.png")
    githubLogo_np = np.frombuffer(githubLogoBytes, np.uint8)
    githubLogo = cv2.imdecode(githubLogo_np, cv2.IMREAD_COLOR)

    qrBytes = generateQR(1, githubUsername)
    qr_np = np.frombuffer(qrBytes, np.uint8)
    qr =  cv2.imdecode(qr_np, cv2.IMREAD_COLOR)

    textDimensions = cv2.getTextSize(githubUsername, cv2.FONT_HERSHEY_DUPLEX, 1.5, 2)
    usernameWidth = textDimensions[0][0]
    xCenter = 550

    #GET TEMPLATES
    if milestone == 1:
        badgeName = 'Bronze'
    elif milestone == 2:
        badgeName = 'Silver'
    elif milestone == 3:
        badgeName = 'Gold'
    else:
        # Extend this logic for ruby and diamond
        print("WEIRD MILESTONE",milestone)
        
    displayTemplateName = f'Github Profile Template {badgeName} Display.png'
    verifiedTemplateName = f'Github Profile Template {badgeName} Verified.png'


    displayTemplateBytes = client.storage.from_('github_profile_assets').download(displayTemplateName)
    displayTemplate_np = np.frombuffer(displayTemplateBytes, np.uint8)
    displayTemplate = cv2.imdecode(displayTemplate_np, cv2.IMREAD_COLOR)

    width, height = displayTemplate.shape[:2]

    MILESTONES_COUNT_COORDINATES = (800, height - 673)
    PRS_RAISED_COORDINATES = (800, height - 608)
    PRS_REVIEWED_COORDINATES = (800, height - 546)
    PRS_MERGED_COORDINATES = (800, height - 483)
    DPG_POINTS_COORDINATES = (800, height - 418)
    QR_COORDINATES = (855, height - 620)
    GITHUB_USERNAME = (xCenter - usernameWidth//2, height - 792)
    GITHUB_LOGO_COORDINATES = (GITHUB_USERNAME[0] - 90, height - 850)

    ghScaleFactor = 0.08
    newGHSize = (int(displayTemplate.shape[1] * ghScaleFactor), int(displayTemplate.shape[1] * ghScaleFactor))
    ghResized = cv2.resize(githubLogo, newGHSize, interpolation=cv2.INTER_AREA)
    displayTemplate[GITHUB_LOGO_COORDINATES[1]:GITHUB_LOGO_COORDINATES[1]+ghResized.shape[0], GITHUB_LOGO_COORDINATES[0]:GITHUB_LOGO_COORDINATES[0]+ghResized.shape[1]] = ghResized

    scaleFactor = 0.107
    newSize = (int(displayTemplate.shape[1] * scaleFactor), int(displayTemplate.shape[1] * scaleFactor))
    qrResized = cv2.resize(qr, newSize, interpolation=cv2.INTER_AREA)
    displayTemplate[QR_COORDINATES[1]:QR_COORDINATES[1]+qrResized.shape[0], QR_COORDINATES[0]:QR_COORDINATES[0]+qrResized.shape[1]] = qrResized



    font = cv2.FONT_HERSHEY_SIMPLEX
    font_size = 1.0
    color = (255, 255, 255)
    thickness = 2

    cv2.putText(displayTemplate, str(milestone), MILESTONES_COUNT_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(prsRaised), PRS_RAISED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(prsReviewed), PRS_REVIEWED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(prsMerged), PRS_MERGED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(dpgPoints), DPG_POINTS_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, f"@{githubUsername}", GITHUB_USERNAME, cv2.FONT_HERSHEY_DUPLEX, 1.5, (216,216,216), thickness)

    verifiedTemplateBytes = client.storage.from_('github_profile_assets').download(verifiedTemplateName)
    verifiedTemplate_np = np.frombuffer(verifiedTemplateBytes, np.uint8)
    verifiedTemplate = cv2.imdecode(verifiedTemplate_np, cv2.IMREAD_COLOR)

    width, height = verifiedTemplate.shape[:2]

    MILESTONES_COUNT_COORDINATES = (800, height - 673)
    PRS_RAISED_COORDINATES = (800, height - 608)
    PRS_REVIEWED_COORDINATES = (800, height - 546)
    PRS_MERGED_COORDINATES = (800, height - 483)
    DPG_POINTS_COORDINATES = (800, height - 418)
    QR_COORDINATES = (855, height - 620)
    GITHUB_USERNAME = (xCenter - usernameWidth//2, height - 792)
    GITHUB_LOGO_COORDINATES = (GITHUB_USERNAME[0] - 90, height - 850)

    ghScaleFactor = 0.08
    newGHSize = (int(verifiedTemplate.shape[1] * ghScaleFactor), int(verifiedTemplate.shape[1] * ghScaleFactor))
    ghResized = cv2.resize(githubLogo, newGHSize, interpolation=cv2.INTER_AREA)
    verifiedTemplate[GITHUB_LOGO_COORDINATES[1]:GITHUB_LOGO_COORDINATES[1]+ghResized.shape[0], GITHUB_LOGO_COORDINATES[0]:GITHUB_LOGO_COORDINATES[0]+ghResized.shape[1]] = ghResized

    scaleFactor = 0.107
    newSize = (int(verifiedTemplate.shape[1] * scaleFactor), int(verifiedTemplate.shape[1] * scaleFactor))
    qrResized = cv2.resize(qr, newSize, interpolation=cv2.INTER_AREA)
    verifiedTemplate[QR_COORDINATES[1]:QR_COORDINATES[1]+qrResized.shape[0], QR_COORDINATES[0]:QR_COORDINATES[0]+qrResized.shape[1]] = qrResized



    font = cv2.FONT_HERSHEY_SIMPLEX
    font_size = 1.0
    color = (255, 255, 255)
    thickness = 2

    cv2.putText(verifiedTemplate, str(milestone), MILESTONES_COUNT_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(prsRaised), PRS_RAISED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(prsReviewed), PRS_REVIEWED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(prsMerged), PRS_MERGED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(dpgPoints), DPG_POINTS_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, f"@{githubUsername}", GITHUB_USERNAME, cv2.FONT_HERSHEY_DUPLEX, 1.5, (216,216,216), thickness)

    display = cv2.imencode(".jpg", displayTemplate)[1].tobytes()
    verified = cv2.imencode(".jpg", verifiedTemplate)[1].tobytes()
    displayFilename = f'{data["discord_id"]}-c4gt-contributions.jpeg'
    verifiedFileName = f'{githubUsername}-c4gt-contributions.jpeg'
    try:
        client.storage.from_("c4gt-github-profile").remove(displayFilename)
        client.storage.from_("c4gt-github-profile").remove(verifiedFileName)
    except Exception:
        pass

    # self.supabase.client.storage.from_(self.storageBucket).remove(displayFilename)
    client.storage.from_("c4gt-github-profile").upload(displayFilename, display, {"content-type": "image/jpeg"})
    client.storage.from_("c4gt-github-profile").upload(verifiedFileName, verified, {"content-type": "image/jpeg"})
```

KDwevedi commented:
Also @sasi2312, I don't see any commits post review, I hope the rest of the feedback on extraneous `pass` statements and more robust error handling can be incorporated.
